internal/provider/models/sources/test/open_telemetry_test.go

```diff
@@ -0,0 +1,193 @@
+			SourceName:    sourceName,
+			SourceTitle:   sourceTitle,
+			SourceDesc:    sourceDesc,
+		}, resourceOpenTelemetrySourceConfigTpl)
+		if err != nil {
+			t.Fatalf("error parsing config template: %s", err)
+		}
+
+		configUpdate, err := ParsedAccConfig(resourceOpenTelemetrySourceConfig{
+			Type:            sourceType,
+			PipelineName:    pipelineName,
+			PipelineTitle:   pipelineTitle,
+			SourceName:      sourceName,
+			SourceTitle:     sourceTitle + " updated",
+			SourceDesc:      sourceDesc,
+			CaptureMetadata: true,
+		}, resourceOpenTelemetrySourceConfigTpl)
+		if err != nil {
+			t.Fatalf("error parsing config template: %s", err)
+		}
+
+		configGateway, err := ParsedAccConfig(resourceOpenTelemetrySourceConfig{
+			Type:              sourceType,
+			PipelineName:      pipelineName,
+			PipelineTitle:     pipelineTitle,
+			SourceName:        sourceName,
+			SourceTitle:       sourceTitle,
+			SourceDesc:        sourceDesc,
+			GatewayResourceID: resourceName + ".gateway_route_id",
+		}, resourceOpenTelemetrySourceConfigTpl)
+		if err != nil {
+			t.Fatalf("error parsing config template: %s", err)
+		}
+
+		configGatewayUpdateErr, err := ParsedAccConfig(resourceOpenTelemetrySourceConfig{
+			Type:              sourceType,
+			PipelineName:      pipelineName,
+			PipelineTitle:     pipelineTitle,
+			SourceName:        sourceName,
+			SourceTitle:       sourceTitle,
+			SourceDesc:        sourceDesc,
+			GatewayResourceID: resourceNamePipeline + ".id",
+		}, resourceOpenTelemetrySourceConfigTpl)
+		if err != nil {
+			t.Fatalf("error parsing config template: %s", err)
+		}
+
+		resource.Test(t, resource.TestCase{
+			ProtoV6ProviderFactories: testAccProtoV6ProviderFactories,
+			PreCheck:                 func() { TestPreCheck(t) },
+			Steps: []resource.TestStep{
+				// Error: pipeline_id is required
+				{
+					Config: fmt.Sprintf(`
+					resource "mezmo_pipeline" "test_parent" {
+						title = "parent pipeline"
+					}
+					resource "mezmo_open_telemetry_%s_source" "my_source" {
+						title = "my kafka title"
+						description = "my kafka description"
+					}`, sourceType),
+					ExpectError: regexp.MustCompile("The argument \"pipeline_id\" is required"),
+				},
+				// Create and Read testing
+				{
+					Config: GetProviderConfig() + config,
+					Check: resource.ComposeTestCheckFunc([]resource.TestCheckFunc{
+						resource.TestMatchResourceAttr(resourceName, "id", IDRegex),
+						resource.TestMatchResourceAttr(resourceName, "gateway_route_id", IDRegex),
+						resource.TestCheckResourceAttr(resourceName, "description", "my open telemetry description"),
+						resource.TestCheckResourceAttr(resourceName, "generation_id", "0"),
+						resource.TestCheckResourceAttr(resourceName, "title", "my open telemetry title"),
+						resource.TestCheckResourceAttr(resourceName, "capture_metadata", "false"),
+						resource.TestCheckResourceAttrPair(resourceName, "pipeline_id", "mezmo_pipeline.test_parent", "id"),
```

Comment:
I see the benefit of this, and like the addition of the assertion comment that `TestCheckResourceAttPair` provides.  I think the initial hangup with this approach was having to duplicate this function call for *every* property tested.  With certain schemas that are large, that makes for a lot of code.  With `StateHasExpectedValues`, the `DEBUG_ATTRIBUTES=1` flag lets it dump what it finds, then the developer can c&p it into a single assertion.

That being said, this is a decent approach for smaller tests and we could even improve on it in the future.